compiler/src/lnntoamm/Stmt.ts

```diff
@@ -0,0 +1,351 @@
+    this.variables = variables !== null ? variables : {};
+  }
+
+  get(name: string): VarDef | null {
+    if (!this.variables.hasOwnProperty(name)) {
+      return null;
+    }
+    return this.variables[name];
+  }
+
+  define(dec: VarDef) {
+    if (this.get(dec.name) !== null) {
+      throw new Error(`Can't redefine value ${dec.name}`);
+    }
+    this.variables[dec.name] = dec;
+  }
+}
+
+export default abstract class Stmt {
+  ast: LPNode
+
+  constructor(
+    ast: LPNode,
+  ) {
+    this.ast = ast;
+  }
+
+  abstract inline(amm: Output): void;
+
+  static fromAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts = [];
+    if (ast.has('assignables')) {
+      let [ss, expr] = Expr.fromAssignablesAst(ast.get('assignments'), metadata);
+      stmts.push(...ss, Dec.gen(expr, metadata));
+    } else if (ast.has('assignments')) {
+      stmts.push(...Assign.fromAssignments(ast.get('assignments'), metadata));
+    } else if (ast.has('conditionals')) {
+      stmts.push(...Cond.fromConditionals(ast.get('conditionals'), metadata));
+    } else if (ast.has('declarations')) {
+      stmts.push(...Dec.fromDeclarations(ast.get('declarations'), metadata));
+    } else if (ast.has('emits')) {
+      stmts.push(...Emit.fromEmits(ast.get('emits'), metadata));
+    } else if (ast.has('exits')) {
+      stmts.push(...Exit.fromExits(ast.get('exits'), metadata));
+    } else {
+      throw new Error(`unrecognized statement ast: ${ast}`);
+    }
+    return stmts;
+  }
+}
+
+class Assign extends Stmt {
+  upstream: VarDef
+  expr: Expr
+
+  constructor(
+    ast: LPNode,
+    upstream: VarDef,
+    expr: Expr,
+  ) {
+    super(ast);
+    this.upstream = upstream;
+    this.expr = expr;
+    if (this.upstream.immutable) {
+      throw new Error(`cannot reassign to a const value`);
+    }
+  }
+
+  static fromAssignments(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts: Stmt[] = [];
+    const name = ast.get('varn').t;
+    const upstream = metadata.get(name);
+    let [generated, expr] = Expr.fromAssignablesAst(ast, metadata);
+    upstream.ty.constrain(expr.ty);
+    stmts.push(...generated, new Assign(ast, upstream, expr));
+    return stmts;
+  }
+
+  inline(amm: Output) {
+    this.expr.inline(amm, '', this.upstream.ammName, this.upstream.ty.breakdown());
+  }
+}
+
+class Cond extends Stmt {
+  static fromConditionals(_ast: LPNode, _metadata: MetaData): Stmt[] {
+    return TODO('conditionals');
+  }
+
+  inline(_amm: Output) {
+    TODO('conditionals');
+  }
+}
+
+export abstract class VarDef extends Stmt {
+  immutable: boolean
+  name: string
+  ty: Type
+
+  abstract get ammName(): string;
+
+  constructor(
+    ast: LPNode,
+    immutable: boolean,
+    name: string,
+    ty: Type,
+  ) {
+    super(ast);
+    this.immutable = immutable;
+    this.name = name;
+    this.ty = ty;
+  }
+
+  ref(): Ref {
+    return new Ref(this);
+  }
+}
+
+export class Dec extends VarDef {
+  private __ammName: string = '';
```

Comment:
A reason for both the `private` and the underscore prefix on this field? I think the underscore prefix convention comes from Python as a way to signal "this is a private field, but you can mess with it if you absolutely have to."